src/main/daml/Daml/Finance/Instrument/Swap/Fpml/Util.daml

```diff
@@ -0,0 +1,111 @@
+-- Copyright (c) 2022 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+module Daml.Finance.Instrument.Swap.Fpml.Util where
+
+import ContingentClaims.Claim (Inequality(..), give, one, scale, when)
+import ContingentClaims.Observation (Observation(..))
+import DA.Date
+import DA.Optional (fromOptional, fromSome, isNone, isSome)
+import Daml.Finance.Instrument.Generic.Util
+import Daml.Finance.Interface.Claims.Types (Deliverable, Observable, TaggedClaim(..))
+import Daml.Finance.Interface.Types.Date.Calendar
+import Daml.Finance.Interface.Types.Date.DayCount (DayCountConventionEnum)
+import Daml.Finance.Interface.Types.Date.RollConvention
+import Daml.Finance.Interface.Types.Date.Schedule (PeriodicSchedule(..), SchedulePeriod, StubPeriodTypeEnum(..))
+import Daml.Finance.Interface.Types.Fpml
+import Daml.Finance.Util.Date.Calendar (addBusinessDays, adjustDate, merge)
+import Daml.Finance.Util.Date.DayCount (calcPeriodDcf)
+import Daml.Finance.Util.Date.RollConvention (addPeriod)
+import Prelude hiding (key)
+
+-- | Create claims from swapStream that describes a fix coupon stream.
+calculateClaimsFromFixSwapStream : SwapStream -> PeriodicSchedule -> [SchedulePeriod] -> Bool -> Bool -> Deliverable -> Update TaggedClaim
+calculateClaimsFromFixSwapStream s periodicSchedule streamSchedule useAdjustedDatesForDcf issuerPaysLeg currency = do
+  let f = fromSome s.calculationPeriodAmount.calculation.fixedRateSchedule
+  pure $ createFixRatePaymentClaims streamSchedule periodicSchedule useAdjustedDatesForDcf f.initialValue issuerPaysLeg s.calculationPeriodAmount.calculation.dayCountFraction currency
+
+-- | Rule when one specific floating rate is provided for a stub period.
+getSingleStubRateFromFpml : DA.Internal.Record.HasField "floatingRateIndex" r o => r -> Optional (Observation t x o)
```

Comment:
Thanks for suggesting `Optional (Observation Time Decimal Observable)`. This is what I was looking for. Is there a systematic way of figuring this out? The compiler suggested `Optional (Observation t x o)` and similar variants. It is a bit difficult to see the proper solution in Daml Studio. 

Also, I noticed that when you just add one function signature, other functions error out until you have added signatures to them as well. Seems manual and compiler-inferred signatures are sometimes not compatible.

By the way, I had to change it to `Optional (Observation Date Decimal Observable)`, otherwise `TimeGte p.adjustedEndDate` gives an error.

Reply:
Thanks! 

Within `Daml Finance` we use `Time` as the time parameter for a claim and `Observable = Text` as the `Observable` parameter. This is specified in https://github.com/digital-asset/daml-finance/blob/main/src/main/daml/Daml/Finance/Interface/Claims/Types.daml#L22.

It probably makes sense to expose a `type O = Observation Date Decimal Observable` in the same file.

Regarding signature inference, I can suggest to write down the signature for your function before you write the function implementation. This is common practice in Haskell and typically at that stage you know exactly what types you plan to work with.